File: sample-conference-videochat/sample-conference-videochat/Components/Users/ViewControllers/UsersViewController/UsersViewController.m
//
//  UsersViewController.m
//  LoginComponent
//
//  Created by Andrey Ivanov on 02/06/16.
//  Copyright © 2016 Quickblox. All rights reserved.
//

#import "UsersViewController.h"
#import <Quickblox/Quickblox.h>
#import "QBCore.h"
#import "UsersDataSource.h"
#import "QBDataFetcher.h"
#import "SVProgressHUD.h"

@implementation UsersViewController

// MARK: Lifecycle

- (void)dealloc {
    
    ILog(@"%@ - %@",  NSStringFromSelector(_cmd), self);
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.tableView.dataSource = self.dataSource;
    self.tableView.rowHeight = 44;
    
    // adding refresh control task
    if (self.refreshControl) {
        
        [self.refreshControl addTarget:self
                                action:@selector(fetchData)
                      forControlEvents:UIControlEventValueChanged];
    }
    
    UIBarButtonItem *createChatButton =
    [[UIBarButtonItem alloc] initWithTitle:@"Create"
                                     style:UIBarButtonItemStylePlain
                                    target:self
                                    action:@selector(didPressCreateChatButton:)];
    
    self.navigationItem.rightBarButtonItem = createChatButton;
    self.navigationItem.rightBarButtonItem.enabled = NO;
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    
    if (self.refreshControl.refreshing) {
        [self.tableView setContentOffset:CGPointMake(0, -self.refreshControl.frame.size.height) animated:NO];
    }
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    
    if ([self isMovingFromParentViewController]) {
        [self.dataSource deselectAllObjects];
    }
}

- (void)didPressCreateChatButton:(UIBarButtonItem *)item {
    
    if ([self hasConnectivity]) {
        
        NSArray *selectedUsers = self.dataSource.selectedObjects;
        NSArray *userIDs = [selectedUsers valueForKeyPath:@"ID"];
        NSArray *userNames = [selectedUsers valueForKey:@"fullName"];
        QBChatDialog *chatDialog = [[QBChatDialog alloc] initWithDialogID:nil type:QBChatDialogTypeGroup];
        chatDialog.occupantIDs = userIDs;
        chatDialog.name = [NSString stringWithFormat:@"%@, %@", Core.currentUser.fullName, [userNames componentsJoinedByString:@", "]];
        
        [SVProgressHUD showWithStatus:NSLocalizedString(@"Creating chat dialog.", nil)];
        __weak __typeof(self)weakSelf = self;
        [QBRequest createDialog:chatDialog successBlock:^(QBResponse * _Nonnull response, QBChatDialog * _Nullable createdDialog) {
            
            [SVProgressHUD dismiss];
            [weakSelf.delegate usersViewController:weakSelf didCreateChatDialog:createdDialog];
            [weakSelf.navigationController popViewControllerAnimated:YES];
            
        } errorBlock:^(QBResponse * _Nonnull response) {
            
            [SVProgressHUD showErrorWithStatus:[NSString stringWithFormat:@"%@", response.error.reasons]];
        }];
    }
}

// MARK: UITableViewDelegate

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    
    [self.dataSource selectObjectAtIndexPath:indexPath];
    [tableView reloadSections:[NSIndexSet indexSetWithIndex:0] withRowAnimation:UITableViewRowAnimationNone];
    
    self.navigationItem.rightBarButtonItem.enabled = self.dataSource.selectedObjects.count > 0;
}

// MARK: Actions

- (BOOL)hasConnectivity {
    
    BOOL hasConnectivity = Core.networkStatus != QBNetworkStatusNotReachable;
    
    if (!hasConnectivity) {
        [self showAlertViewWithMessage:NSLocalizedString(@"Please check your Internet connection", nil)];
    }
    
    return hasConnectivity;
}

- (void)showAlertViewWithMessage:(NSString *)message {
    
    UIAlertController *alertController =
    [UIAlertController alertControllerWithTitle:nil
                                        message:message
                                 preferredStyle:UIAlertControllerStyleAlert];
    
    [alertController addAction:[UIAlertAction actionWithTitle:NSLocalizedString(@"Ok", nil)
                                                        style:UIAlertActionStyleDefault
                                                      handler:nil]];
    
    [self presentViewController:alertController animated:YES completion:nil];
}

// MARK: Private

- (void)fetchData {
    
    __weak __typeof(self)weakSelf = self;
    [QBDataFetcher fetchUsers:^(NSArray *users) {
        
        [weakSelf.dataSource setObjects:users];
        [weakSelf.tableView reloadData];
        [weakSelf.refreshControl endRefreshing];
    }];
}

@end

File: sample-conference-videochat-swift/sample-conference-videochat-swift/sample-conference-videochat-swift/Components/Users/ViewControllers/UsersViewController/UsersViewController.swift
//
//  UsersViewController.swift
//  sample-conference-videochat-swift
//
//  Created by Vladimir Nybozhinsky on 12.10.2018.
//  Copyright © 2018 QuickBlox. All rights reserved.
//

import UIKit
import Quickblox
import SVProgressHUD

protocol UsersViewControllerDelegate: class {
    func usersViewController(_ usersViewController: UsersViewController,
                             didCreateChatDialog chatDialog: QBChatDialog?)
}

struct UsersViewControllerConstant {
    static let create = "Create"
    static let creatingChatDialog = NSLocalizedString("Creating chat dialog.", comment: "")
    static let checkInternetConnection = NSLocalizedString("Please check your Internet connection",
                                                           comment: "")
}

class UsersViewController: UITableViewController {
    // MARK: - Properties
    let core = Core.instance
    weak var dataSource: UsersDataSource?
    weak var delegate: UsersViewControllerDelegate?
    
    // MARK: Life Сycle
    override func viewDidLoad() {
        super.viewDidLoad()
        tableView.dataSource = dataSource
        tableView.rowHeight = 44.0
        fetchData()
        
        // adding refresh control task
        if (refreshControl != nil) {
            refreshControl?.addTarget(self, action: #selector(fetchData),
                                      for: .valueChanged)
        }
        
        let createChatButton = UIBarButtonItem(title: UsersViewControllerConstant.create,
                                               style: .plain,
                                               target: self,
                                               action: #selector(didTapCreateChatButton(_:)))
        
        navigationItem.rightBarButtonItem = createChatButton
        navigationItem.rightBarButtonItem?.isEnabled = false
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        if let refreshControl = self.refreshControl, refreshControl.isRefreshing == true {
            let contentOffset = CGPoint(x: 0, y: -refreshControl.frame.size.height)
            tableView.setContentOffset(contentOffset, animated: false)
        }
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        if isMovingFromParent {
            dataSource?.deselectAllObjects()
        }
    }
    
    deinit {
        debugPrint("deinit \(self)")
    }
    
    //MARK: - Actions
    @objc private func didTapCreateChatButton(_ item: UIBarButtonItem?) {
        if hasConnectivity() == true {
            
            let selectedUsers = dataSource?.selectedObjects
            let userIDs = selectedUsers?.map{ $0.id }
            let userNames = selectedUsers?.map{ $0.fullName } as! [String]
            let chatDialog = QBChatDialog(dialogID: nil, type: QBChatDialogType.group)
            chatDialog.occupantIDs = userIDs as [NSNumber]?
            
            guard let fullName = core.currentUser?.fullName else { return }
            let userNamesString = userNames.joined(separator: ", ")
            chatDialog.name = "\(fullName), \(userNamesString)"
            
            SVProgressHUD.show(withStatus: UsersViewControllerConstant.creatingChatDialog)
            
            QBRequest.createDialog(chatDialog, successBlock: { [weak self] response, createdDialog in
                guard let `self` = self else { return }
                SVProgressHUD.dismiss()
                self.delegate?.usersViewController(self, didCreateChatDialog: createdDialog)
                self.navigationController?.popViewController(animated: true)
                
                }, errorBlock: { response in
                    SVProgressHUD.showError(withStatus: "\(String(describing: response.error?.reasons))")
            })
        }
    }
    @objc private func fetchData() {
        QBDataFetcher.fetchUsers({ [weak self] users in
            if let users = users {
                self?.dataSource?.updateObjects(users)
                self?.tableView.reloadData()
                self?.refreshControl?.endRefreshing()
            } else {
                self?.showAlertView(withMessage:  UsersViewControllerConstant.checkInternetConnection)
            }
        })
    }
    
    // MARK: UITableViewDelegate
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        dataSource?.selectObject(at: indexPath)
        var countUsers = 0
       if let count = dataSource?.selectedObjects.count {
           countUsers = count
        }
        tableView.reloadSections(NSIndexSet(index: 0) as IndexSet, with: .none)
        navigationItem.rightBarButtonItem?.isEnabled = countUsers > 0
    }
    
    // MARK: - Internal Methods
    private func hasConnectivity() -> Bool {
        let hasConnectivity = core.networkConnectionStatus() != NetworkConnectionStatus.notConnection
        if hasConnectivity == false {
            showAlertView(withMessage: UsersViewControllerConstant.checkInternetConnection)
        }
        return hasConnectivity
    }
    
    private func showAlertView(withMessage message: String?) {
        let alertController = UIAlertController(title: nil, message: message, preferredStyle: .alert)
        alertController.addAction(UIAlertAction(title: NSLocalizedString("Ok", comment: ""),
                                                style: .default, handler: nil))
        present(alertController, animated: true)
    }
}
